### PR TITLE
Add manufacturers #68

### LINKED ISSUE #68: As an operator/admin, I want to be able to add manufacturers
- [ ] Add a new Manufacturer
    -  [ ] Implement API post request
    -  [ ] Create a UI design (add catalogue items window)
    -  [ ] Implement the UI design in the add catalogue items window 
    -  [ ] Implement the UI design in the Manufacturer tab (UI design similar to add catalogue items)
    -  [ ] Implement error checking

### PR DESCRIPTION
**Pull request by MatteoGuarnaccia5**
## Description

Add a manufacturer via a text form, all the fields need to be filled and there is relevant error checks (for invalid urls and duplicate names as well)

## Testing instructions

Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] test adding and also using invalid url, empty strings(spaces) =

## Agile board tracking

closes #68

### CODE REVIEW

**Review by joelvdavies — CHANGES_REQUESTED** [review]
A couple of things I noticed - With the exception of the missing grid elements (and console.log), I am happy for the other changes to be addressed in #112 if you prefer.

File: src/api/manufacturer.tsx [hunk 1]
```diff
@@ -33,3 +43,36 @@ export const useManufacturers = (): UseQueryResult<
...
+  const settingsResult = await settings;
+  if (settingsResult) {
+    apiUrl = settingsResult['apiUrl'];
+  }
+  return axios
+    .post<AddManufacturerResponse>(`${apiUrl}/v1/manufacturers`, manufacturer)
+    .then((response) => response.data);
+};
+
+export const useAddManufacturer = (): UseMutationResult<
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/api/manufacturer.tsx:61]: Addressed in https://github.com/ral-facilities/inventory-management-system/pull/97

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    addressBuildingNumberErrorMessage,
+    setAddressBuildingNumberErrorMessage,
+  ] = React.useState<string | undefined>(undefined);
+  const [addressStreetNameError, setAddressStreetNameError] =
+    React.useState(false);
+  const [addressStreetNameErrorMessage, setaddressStreetNameErrorMessage] =
+    React.useState<string | undefined>(undefined);
+  const [addresspostcodeError, setAddresspostcodeError] = React.useState(false);
+  const [AddresspostcodeErrorMessage, setAddresspostcodeErrorMessage] =
+    React.useState<string | undefined>(undefined);
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/manufacturer/manufacturerDialog.component.tsx:60]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/112

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+function AddManufacturerDialog(props: AddManufacturerDialogProps) {
+  const { open, onClose, manufacturer, onChangeManufacturerDetails } = props;
+
+  const [nameError, setNameError] = React.useState(false);
+  const [nameErrorMessage, setNameErrorMessage] = React.useState<
+    string | undefined
+  >(undefined);
+  const [URlerror, setURLError] = React.useState(false);
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/manufacturer/manufacturerDialog.component.tsx:44]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/112

File: src/app.types.tsx [hunk 4]
```diff
@@ -36,10 +36,34 @@ export interface CatalogueCategory {
...
+}
+
+export interface AddManufacturerResponse {
+  name: string;
+  code: string;
+  url: string;
+  address: Address;
+  telephone: string;
+  id: string;
+}
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/app.types.tsx:58]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/112

File: src/app.types.tsx [hunk 5]
```diff
@@ -96,6 +120,13 @@ export interface ErrorParsing {
   detail: string;
 }
 
+interface Address {
+  building_number: string;
+  street_name: string;
+  town?: string;
+  county?: string;
+  postcode: string;
+}
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/app.types.tsx:129]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/118

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    setURLErrorMessage(undefined);
+    setAddressBuildingNumberError(false);
+    setAddressBuildingNumberErrorMessage(undefined);
+    setAddressStreetNameError(false);
+    setaddressStreetNameErrorMessage(undefined);
+    setAddresspostcodeError(false);
+    setAddresspostcodeErrorMessage(undefined);
+    onClose();
+  }, [onClose, onChangeManufacturerDetails]);
+
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/manufacturer/manufacturerDialog.component.tsx:89]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/112

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+      url: manufacturer.url,
+      address: {
+        building_number: manufacturer.address.building_number,
+        street_name: manufacturer.address.street_name,
+        town: manufacturer.address.town,
+        county: manufacturer.address.county,
+        postcode: manufacturer.address.postcode,
+      },
+      telephone: manufacturer.telephone,
+    };
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/manufacturer/manufacturerDialog.component.tsx:151]: Should be addressed in https://github.com/ral-facilities/inventory-management-system/pull/112

**Review by joelvdavies — APPROVED** [review]
Should be merged with https://github.com/ral-facilities/inventory-management-system/pull/80.

File: src/manufacturer/manufacturerDialog.component.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/manufacturer/manufacturerDialog.component.tsx:44]: addressed in #112 as 'add manufacturer changes' commit

File: src/manufacturer/manufacturerDialog.component.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/manufacturer/manufacturerDialog.component.tsx:60]: addressed in #112 as 'add manufacturer changes' commit

File: src/manufacturer/manufacturerDialog.component.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/manufacturer/manufacturerDialog.component.tsx:151]: addressed in #112 as 'add manufacturer changes' commit

File: src/manufacturer/manufacturerDialog.component.tsx [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+  }
+}
+
+function AddManufacturerDialog(props: AddManufacturerDialogProps) {
+  const { open, onClose, manufacturer, onChangeManufacturerDetails } = props;
+
+  const [nameError, setNameError] = React.useState(false);
+  const [nameErrorMessage, setNameErrorMessage] = React.useState<
+    string | undefined
+  >(undefined);
```
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply src/manufacturer/manufacturerDialog.component.tsx:43]: That's a good suggestion. Could you create an issue so we can track these changes please. I'm going to change this in my updated catalogue items fields pr

[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/manufacturer/manufacturerDialog.component.tsx:43]: Addressed in #118 PR with commit name 'change errors to use less states'

File: src/app.types.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/app.types.tsx:58]: addressed in #118 with commit name 'changed types to be | null'

File: src/app.types.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (MatteoGuarnaccia5) [reply src/app.types.tsx:129]: addressed in https://github.com/ral-facilities/inventory-management-system/pull/118 with commit name 'changed types to be | null'

**Pull request merged**